api/src/main/java/io/grpc/StatusOr.java

```diff
@@ -0,0 +1,40 @@
+package io.grpc;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/** Either a Status or a value. */
+public class StatusOr<T> {
+  private StatusOr() {}
+
+  public static <T> StatusOr<T> fromValue(T value) {
+    StatusOr<T> result = new StatusOr<T>();
+    result.hasValue = true;
+    result.value = checkNotNull(value, "value");
+    return result;
+  }
+
+  public static <T> StatusOr<T> fromStatus(Status status) {
+    StatusOr<T> result = new StatusOr<T>();
+    result.status  = checkNotNull(status, "status");
+    checkArgument(!status.isOk(), "cannot use OK status: %s", status);
+    result.hasValue = false;
+    return result;
+  }
+
+  public boolean hasValue() {
+    return hasValue;
+  }
+
+  public T value() {
+    return value;
+  }
+
+  public Status status() {
+    return status;
+  }
+
+  private Status status;
```

Comment:
Make these final. And yes, that means you need to pass them to a constructor.

Final is special in Java for synchronization. It allows the object to be used with unsafe publication. We don't really do that, but it is a helpful safety guarantee for immutable objects. For example, CallOptions was swapped to using final in https://github.com/grpc/grpc-java/commit/1b94f48dc84f91977b1553a33e8eccbfaffbfb05

Reply:
Done.